File: army/src/main/java/com/rolan/army/RequestBuilder.java
package com.rolan.army;

import android.content.Context;

import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Rolan.Wang on 1/28/21.6:43 PM
 * 描述：
 */
public class RequestBuilder extends BaseRequestOptions<RequestBuilder>{
    private final Context context;
    private final RequestManager requestManager;
    private final Army army;
    private final ArmyContext armyContext;

    private List<RequestListener> requestListeners;

    public RequestBuilder(Army army, RequestManager requestManager,Context context ) {
        this.context = context;
        this.requestManager = requestManager;
        this.army = army;
        this.armyContext=army.getArmyContext();
        initRequestListeners(requestManager.getDefaultRequestListeners());
    }

    private void initRequestListeners(List<RequestListener> requestListeners) {
        for (RequestListener listener : requestListeners) {
            addListener((RequestListener) listener);
        }
    }

    private Request buildRequest(){
        return SingleRequest.obtain(context,armyContext.getEngine(),armyContext,requestListeners,this);
    }



    public void post(){
        Request request = buildRequest();
        requestManager.track(request);
    }


    private RequestBuilder addListener(
            @Nullable RequestListener requestListener) {
        if (requestListener != null) {
            if (this.requestListeners == null) {
                this.requestListeners = new ArrayList<>();
            }
            this.requestListeners.add(requestListener);
        }
        return this;
    }

    public RequestBuilder listener(
            @Nullable RequestListener requestListener) {
        this.requestListeners = null;
        return addListener(requestListener);
    }
}

File: army/src/main/java/com/rolan/army/RequestBuilder.kt
package com.rolan.army

import android.content.Context
import java.util.*

/**
 * Created by Rolan.Wang on 1/28/21.6:43 PM
 * request builder
 */
class RequestBuilder(private val army: Army, private val requestManager: RequestManager, private val context: Context) : BaseRequestOptions<RequestBuilder>() {
    private val armyContext: ArmyContext = army.armyContext
    private var requestListeners: MutableList<RequestListener>? = null
    private fun initRequestListeners(requestListeners: List<RequestListener>) {
        for (listener in requestListeners) {
            addListener(listener)
        }
    }

    private fun buildRequest(): Request {
        return SingleRequest.obtain(context, armyContext.engine, armyContext, requestListeners, this)
    }

    fun post() {
        val request = buildRequest()
        requestManager.track(request)
    }

    private fun addListener(
            requestListener: RequestListener?): RequestBuilder {
        if (requestListener != null) {
            if (requestListeners == null) {
                requestListeners = ArrayList()
            }
            requestListeners?.add(requestListener)
        }
        return this
    }

    fun listener(
            requestListener: RequestListener?): RequestBuilder {
        return addListener(requestListener)
    }

    init {
        initRequestListeners(requestManager.defaultRequestListeners)
    }
}